### PR TITLE
[SelectionDAG] Add space-optimized forms of OPC_CheckPatternPredicate

### PR DESCRIPTION
**Pull request by wangpc-pp**
We record the usage of each `PatternPredicate` and sort them by
usage.

For the top 8 `PatternPredicate`s, we will emit a
`OPC_CheckPatternPredicateN` to save one byte.

The old `OPC_CheckPatternPredicate2` is renamed to
`OPC_CheckPatternPredicateTwoByte`.

Overall this reduces the llc binary size with all in-tree targets by
about 93K.

This PR is stacked on #73310.

### CODE REVIEW

**Comment by wangpc-pp** [comment]
Ping.

**Comment by wangpc-pp** [comment]
Ping.

**Comment by wangpc-pp** [comment]
Ping.

File: llvm/utils/TableGen/DAGISelMatcherEmitter.cpp [hunk 1]
```diff
@@ -85,7 +83,25 @@ class MatcherTableEmitter {
...
 public:
   MatcherTableEmitter(const CodeGenDAGPatterns &cgp)
-      : CGP(cgp), OpcodeCounts(Matcher::HighestKind + 1, 0) {}
+      : CGP(cgp), OpcodeCounts(Matcher::HighestKind + 1, 0) {
+    // Sort ComplexPatterns by usage.
+    auto &ComplexPatternUsage = cgp.getComplexPatternUsage();
+    std::vector<std::pair<const ComplexPattern *, unsigned>> ComplexPatternList(
+        ComplexPatternUsage.begin(), ComplexPatternUsage.end());
+    sort(ComplexPatternList,
+         [](auto &A, auto &B) { return A.second > B.second; });
```
Comment (arsenm) [inline llvm/utils/TableGen/DAGISelMatcherEmitter.cpp:92]: const?

File: llvm/utils/TableGen/DAGISelMatcherEmitter.cpp [hunk 2]
```diff
@@ -85,7 +83,25 @@ class MatcherTableEmitter {
...
+    for (auto &ComplexPattern : ComplexPatternList)
```
Comment (arsenm) [inline llvm/utils/TableGen/DAGISelMatcherEmitter.cpp:93]: const?

File: llvm/utils/TableGen/DAGISelMatcherGen.cpp [hunk 3]
```diff
@@ -573,8 +572,10 @@ bool MatcherGen::EmitMatcherCode(unsigned Variant) {
   // If the pattern has a predicate on it (e.g. only enabled when a subtarget
   // feature is around, do the check).
   std::string PredicateCheck = Pattern.getPredicateCheck();
-  if (!PredicateCheck.empty())
+  if (!PredicateCheck.empty()) {
+    CGP.increasePatternPredicateUsage(PredicateCheck);
```
Comment (arsenm) [inline llvm/utils/TableGen/DAGISelMatcherGen.cpp:576]: Seems unrelated?

File: llvm/utils/TableGen/DAGISelMatcherGen.cpp [hunk 4]
```diff
@@ -602,6 +603,7 @@ bool MatcherGen::EmitMatcherCode(unsigned Variant) {
 
     // Emit a CheckComplexPat operation, which does the match (aborting if it
     // fails) and pushes the matched operands onto the recorded nodes list.
+    CGP.increaseComplexPatternUsage(CP);
```
Comment (arsenm) [inline llvm/utils/TableGen/DAGISelMatcherGen.cpp:606]: Seems unrelated?

Comment (wangpc-pp) [reply llvm/utils/TableGen/DAGISelMatcherGen.cpp:606]: This is from parent commit, not this PR.

File: llvm/utils/TableGen/DAGISelMatcherGen.cpp (hunk 3 above)
Comment (wangpc-pp) [reply llvm/utils/TableGen/DAGISelMatcherGen.cpp:576]: No, it's related. It's for increasing the usage of `PatternPredicate`.

File: llvm/utils/TableGen/DAGISelMatcherEmitter.cpp (hunk 1 above)
Comment (wangpc-pp) [reply llvm/utils/TableGen/DAGISelMatcherEmitter.cpp:92]: Fixed in #73310.

File: llvm/utils/TableGen/DAGISelMatcherEmitter.cpp (hunk 2 above)
Comment (wangpc-pp) [reply llvm/utils/TableGen/DAGISelMatcherEmitter.cpp:93]: Fixed in #73310.

**Review by arsenm — APPROVED** [review]
